snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,288 @@
+from typing import Optional, Sequence, Set, cast
+
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.query.conditions import (
+    BooleanFunctions,
+    ConditionFunctions,
+    get_first_level_conditions,
+    in_condition,
+    is_binary_condition,
+    is_in_condition_pattern,
+)
+from snuba.query.dsl import arrayElement
+from snuba.query.expressions import Argument
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression
+from snuba.query.expressions import FunctionCall as FunctionCallExpr
+from snuba.query.expressions import Lambda
+from snuba.query.expressions import Literal as LiteralExpr
+from snuba.query.matchers import Any, Column, FunctionCall, Literal, Or, Param, String
+from snuba.request.request_settings import RequestSettings
+
+tags_key_pattern = FunctionCall(
+    None, String("arrayJoin"), (Column(None, None, String("tags.key")),)
+)
+
+
+def _get_tag_keys_in_condition(condition: Optional[Expression]) -> Optional[Set[str]]:
```

Comment:
Does the same as this:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L160

I decided not to try to make the old logic parametric and to just start fresh instead because, specifically in the query processor code the structure is fairly different because we have the matcher api right now.